### PR TITLE
[processing][FEATURE] add multi-ring buffer algorithm

### PR DESCRIPTION
**Pull request by alexbruy**
## Description
Add native multi-ring buffer algorithm with user-defined number of rings and distance between them. Produces "donut"-like buffers around input features, first ring is a normal buffer.
multiring

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nirvn** [comment]
Nice

File: src/analysis/processing/qgsalgorithmmultiringbuffer.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+QgsFeatureList QgsMultiRingBufferAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &, QgsProcessingFeedback *feedback )
+{
+  double currentDistance = 0;
+  QgsGeometry outputGeometry, previousGeometry;
+
+  QgsFeatureList outputs;
+
+  for ( int i = 1; i <= mRingsNumber; ++i )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiringbuffer.cpp]: There's a missing bit in here - we need to evaluate the dynamic parameters here for this feature.

So something like:

    int ringsNumber  = mRingsNumber;
    if ( mDynamicRingsNumber )
        ringsNumber = mRingsNumberProperty.valueAsInt( context.expressionContext(), ringsNumber );

    for ( int i = 1; i <= ringsNumber; ++i )

A similar change is required for mDistance - we'll need to evalute that once outside of the loop and then use the evaluated value inside the loop

**Comment by nyalldawson** [comment]
Sweet - nice to see this inbuilt!

**Comment by nyalldawson** [comment]
Oh one thought... I wonder if we should name this algorithm something like "Multi-ring buffer (constant distance)"? There's a good chance at some stage in the future there'll be an algorithm which accepts a list of preset distances (eg "50,100,200,500"), so if we give this one a less-generic name it'll be easier to add that algorithm at a later stage without changing any existing names.

**Comment by NathanW2** [comment]
Fancy!

File: src/analysis/processing/qgsalgorithmmultiringconstantbuffer.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  if ( mDynamicRingsNumber )
+    rings = mRingsNumberProperty.valueAsInt( context.expressionContext(), rings );
+
+  double distance = mDistance;
+  if ( mDynamicDistance )
+    distance = mDistanceProperty.valueAsDouble( context.expressionContext(), distance );
+
+  QgsFeatureList outputs;
+
+  for ( int i = 1; i <= mRingsNumber; ++i )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmmultiringconstantbuffer.cpp]: shouldn't this be rings, not mRingsNumber?

Comment (alexbruy) [reply src/analysis/processing/qgsalgorithmmultiringconstantbuffer.cpp]: Oops... fixed

**Pull request merged**